File: src/core/system/utils/paths.rb

require 'core/semantics/code/interpreter'
require 'core/schema/tools/equals'
# TODO: error handling

module Paths
  def self.new(start = nil)
    Path.new(start)
  end

  class Path
    include Interpreter::Dispatcher
    
    attr_reader :path

    #[JS HACK] @@ in class methods confuses JS
    def self.set_factory(factory)
      Path.new(factory.EVar("root")).set_factory(factory)
    end
    def set_factory(factory)
      @@factory = factory
    end
    #[/JS HACK] 

    def initialize(path = @@factory.EVar("root"))
      @path = path ? path : @@factory.EVar("root")
    end

    def field(name)
      Path.new(@@factory.EField(@path, name))
    end
    
    def key(key)
      Path.new(@@factory.ESubscript(@path, @@factory.EStrConst(key)))
    end

    def index(index)
      Path.new(@@factory.ESubscript(@path, @@factory.EIntConst(index)))
    end

    def equals(other)
      Equals::equals(@path, other.path)
    end

    def deref?(root)
      begin
        !deref(root).nil?
      rescue
        false
      end
    end
    
    def to_s()
      to_s_path(@path)
    end
    
    def to_s_path(path)
      dispatch_obj(:to_s, path)
    end
    
    def to_s_EVar(obj)
      obj.name
    end

    def to_s_EConst(obj)
      obj.val
    end
        
    def to_s_EField(obj)
      "#{to_s_path(obj.e)}.#{obj.fname}"
    end

    def to_s_ESubscript(obj)
      "#{to_s_path(obj.e)}[#{to_s_path(obj.sub)}]"
    end

    def deref(root)
      dynamic_bind(root: root) do 
        eval
      end
    end

    def eval(path = @path)
      dispatch_obj(:eval, path)
    end

    def eval_EVar(obj)
      raise "undefined variable #{obj.name}" if !@D.include?(obj.name.to_sym)
      #puts("VAR #{obj.name} => #{@D[obj.name.to_sym]}")
      @D[obj.name.to_sym]
    end

    def eval_EConst(obj)
      obj.val
    end

    def eval_EField(obj)
      eval(obj.e)[obj.fname]
    end

    def eval_ESubscript(obj)
      eval(obj.e)[eval(obj.sub)]
    end

    def assign(root, val)
      obj = @path
      if obj.EField?
        dynamic_bind(root: root) do
          eval(obj.e)[obj.fname] = val
        end
      elsif obj.ESubscript?
        dynamic_bind(root: root) do 
          eval(obj.e)[eval(obj.sub)] = val
        end
      end
    end

    #insert is the same as assign, except that for arrays it injects into rather than replace
    def insert(root, val)
      obj = @path
      if obj.EField?
        dynamic_bind(root: root) do
          eval(obj.e)[obj.fname] = val
        end
      elsif obj.ESubscript?
        dynamic_bind(root: root) do 
          eval(obj.e).insert(eval(obj.sub), val)
        end
      end
    end
    
    def delete(root)
      obj = @path
      if obj.EField?
        dynamic_bind(root: root) do
          eval(obj.e)[obj.fname] = nil
        end
      elsif obj.ESubscript?
        dynamic_bind(root: root) do
          eval(obj.e).delete(eval(obj))
        end
      end
    end

    def type(root, obj = @path)
      if obj.EField?
        dynamic_bind(root: root) do
          eval(obj.e).schema_class.fields[obj.fname]
        end
      elsif obj.ESubscript?
        type(root, obj.e)
      end
    end

    def assign_and_coerce(root, value)
      raise "Can only assign to lvalues not to #{self}" if not lvalue?
      obj = owner.deref(root)
      fld = obj.schema_class.fields[last.name]
      if fld.type.Primitive? then
          case fld.type.name 
          when 'str' then value = value.to_s
          when 'int' then value = value.to_i
          when 'bool' then value = (value.to_s == 'true') ? true : false
          when 'real' then value = value.to_f
          else
            raise "Unknown primitive type: #{fld.type.name}"
          end
      end
      owner.deref(root)[last.name] = value
    end
  end
end

File: src/js/core/system/utils/paths.js
define(["core/semantics/code/interpreter", "core/schema/tools/equals"], (function (Interpreter, Equals) {
  var Paths;
  var Path = MakeClass("Path", null, [Interpreter.Dispatcher], (function () {
    (this.set_factory = (function (factory) {
      var self = this;
      return Path.new(factory.EVar("root")).set_factory(factory);
    }));
  }), (function (super$) {
    (this.path = (function () {
      return this.$.path;
    }));
    (this.set_factory = (function (factory) {
      var self = this;
      return (self._class_.$.factory = factory);
    }));
    (this.initialize = (function (path) {
      var self = this;
      (path = (((typeof path) !== "undefined") ? path : self._class_.$.factory.EVar("root")));
      return (self.$.path = (path ? path : self._class_.$.factory.EVar("root")));
    }));
    (this.field = (function (name) {
      var self = this;
      return Path.new(self._class_.$.factory.EField(self.$.path, name));
    }));
    (this.key = (function (key) {
      var self = this;
      return Path.new(self._class_.$.factory.ESubscript(self.$.path, self._class_.$.factory.EStrConst(key)));
    }));
    (this.index = (function (index) {
      var self = this;
      return Path.new(self._class_.$.factory.ESubscript(self.$.path, self._class_.$.factory.EIntConst(index)));
    }));
    (this.equals = (function (other) {
      var self = this;
      return Equals.equals(self.$.path, other.path());
    }));
    (this.deref_P = (function (root) {
      var self = this;
      try {return (!(self.deref(root) == null));
           
      }
      catch (caught$960) {
        
          return false;
      }
    }));
    (this.to_s = (function () {
      var self = this;
      return self.to_s_path(self.$.path);
    }));
    (this.to_s_path = (function (path) {
      var self = this;
      return self.dispatch_obj("to_s", path);
    }));
    (this.to_s_EVar = (function (obj) {
      var self = this;
      return obj.name();
    }));
    (this.to_s_EConst = (function (obj) {
      var self = this;
      return obj.val();
    }));
    (this.to_s_EField = (function (obj) {
      var self = this;
      return S("", self.to_s_path(obj.e()), ".", obj.fname(), "");
    }));
    (this.to_s_ESubscript = (function (obj) {
      var self = this;
      return S("", self.to_s_path(obj.e()), "[", self.to_s_path(obj.sub()), "]");
    }));
    (this.deref = (function (root) {
      var self = this;
      return self.dynamic_bind((function () {
        return self.eval();
      }), (new EnsoHash({
        root: root
      })));
    }));
    (this.eval = (function (path) {
      var self = this;
      (path = (((typeof path) !== "undefined") ? path : self.$.path));
      return self.dispatch_obj("eval", path);
    }));
    (this.eval_EVar = (function (obj) {
      var self = this;
      if ((!self.$.D.include_P(obj.name().to_sym()))) {
        self.raise(S("undefined variable ", obj.name(), ""));
      }
      return self.$.D._get(obj.name().to_sym());
    }));
    (this.eval_EConst = (function (obj) {
      var self = this;
      return obj.val();
    }));
    (this.eval_EField = (function (obj) {
      var self = this;
      return self.eval(obj.e())._get(obj.fname());
    }));
    (this.eval_ESubscript = (function (obj) {
      var self = this;
      return self.eval(obj.e())._get(self.eval(obj.sub()));
    }));
    (this.assign = (function (root, val) {
      var self = this;
      var obj;
      (obj = self.$.path);
      if (obj.EField_P()) { 
        return self.dynamic_bind((function () {
          return self.eval(obj.e())._set(obj.fname(), val);
        }), (new EnsoHash({
          root: root
        }))); 
      }
      else { 
        if (obj.ESubscript_P()) { 
          return self.dynamic_bind((function () {
            return self.eval(obj.e())._set(self.eval(obj.sub()), val);
          }), (new EnsoHash({
            root: root
          }))); 
        } 
        else {
             }
      }
    }));
    (this.insert = (function (root, val) {
      var self = this;
      var obj;
      (obj = self.$.path);
      if (obj.EField_P()) { 
        return self.dynamic_bind((function () {
          return self.eval(obj.e())._set(obj.fname(), val);
        }), (new EnsoHash({
          root: root
        }))); 
      }
      else { 
        if (obj.ESubscript_P()) { 
          return self.dynamic_bind((function () {
            return self.eval(obj.e()).insert(self.eval(obj.sub()), val);
          }), (new EnsoHash({
            root: root
          }))); 
        } 
        else {
             }
      }
    }));
    (this.delete = (function (root) {
      var self = this;
      var obj;
      (obj = self.$.path);
      if (obj.EField_P()) { 
        return self.dynamic_bind((function () {
          return self.eval(obj.e())._set(obj.fname(), null);
        }), (new EnsoHash({
          root: root
        }))); 
      }
      else { 
        if (obj.ESubscript_P()) { 
          return self.dynamic_bind((function () {
            return self.eval(obj.e()).delete(self.eval(obj));
          }), (new EnsoHash({
            root: root
          }))); 
        } 
        else {
             }
      }
    }));
    (this.type = (function (root, obj) {
      var self = this;
      (obj = (((typeof obj) !== "undefined") ? obj : self.$.path));
      if (obj.EField_P()) { 
        return self.dynamic_bind((function () {
          return self.eval(obj.e()).schema_class().fields()._get(obj.fname());
        }), (new EnsoHash({
          root: root
        }))); 
      }
      else { 
        if (obj.ESubscript_P()) { 
          return self.type(root, obj.e()); 
        } 
        else {
             }
      }
    }));
    (this.assign_and_coerce = (function (root, value) {
      var self = this;
      var obj, fld;
      if ((!self.lvalue_P())) {
        self.raise(S("Can only assign to lvalues not to ", self, ""));
      }
      (obj = self.owner().deref(root));
      (fld = obj.schema_class().fields()._get(self.last().name()));
      if (fld.type().Primitive_P()) {
        switch ((function () {
          return fld.type().name();
        })()) {
          case "real":
           (value = value.to_f());
           break;
          case "bool":
           (value = ((value.to_s() == "true") ? true : false));
           break;
          case "int":
           (value = value.to_i());
           break;
          case "str":
           (value = value.to_s());
           break;
          default:
           self.raise(S("Unknown primitive type: ", fld.type().name(), ""));
        }
            
      }
      return self.owner().deref(root)._set(self.last().name(), value);
    }));
  }));
  (Paths = {
    new: (function (start) {
      var self = this;
      (start = (((typeof start) !== "undefined") ? start : null));
      return Path.new(start);
    }),
    Path: Path
  });
  return Paths;
}));